### PR TITLE
Handle the case of remote users leaving a partial join room for device lists

### PR DESCRIPTION
**Pull request by erikjohnston**
This is bullet two in https://github.com/matrix-org/synapse/issues/12993#issuecomment-1253785129

### CODE REVIEW

File: synapse/storage/controllers/persist_events.py [hunk 1]
```diff
@@ -1195,45 +1165,4 @@ async def _is_server_still_joined(
...
-                state_key,
-            ), event_id in current_state.items()
-            if typ == EventTypes.Member and not self.is_mine_id(state_key)
-        ]
-        members = await self.main_store.get_membership_from_event_ids(remote_event_ids)
-        potentially_left_users.update(
-            member.user_id
-            for member in members.values()
-            if member and member.membership == Membership.JOIN
-        )
```
Comment (erikjohnston) [inline synapse/storage/controllers/persist_events.py:1221]: All this logic has moved into `events.py`.

File: synapse/storage/databases/main/events.py [hunk 2]
```diff
@@ -1183,6 +1183,12 @@ def _update_current_state_txn(
                 txn, room_id, members_changed
             )
 
+            # Check if any of the remote membership changes requires us to
+            # unsubscribe from their device lists.
+            self.store.handle_potentially_left_users_txn(
+                txn, {m for m in members_changed if not self.hs.is_mine_id(m)}
+            )
```
Comment (erikjohnston) [inline synapse/storage/databases/main/events.py:1190]: Oh, we no longer check if the old memberships were joins, which makes this slightly more expensive than at present.

**Comment by squahtx** [comment]
Thanks for taking this on!

I've got a WIP complement branch with the relevant tests at `squah/faster_room_joins_device_list_tracking` and they pass (see the `TestPartialStateJoin/Device_list_tracking/.*incorrect*` tests).

**Review by squahtx — APPROVED** [review]
lgtm!

**Comment by richvdh** [comment]
To confirm I understand: this is fixing a bug where:
 * we thought we were sharing a (partial-state) room with a user (`current_state_events` includes a membership for that user)
 * during the state resync, we actually decided that user was not in the room (so the membership is removed from `current_state_events`)
 * which left us in the state where we thought we were sharing a room with a given user, but we were not.

?

**Comment by erikjohnston** [comment]
Yup, this is bullet two in this comment: https://github.com/matrix-org/synapse/issues/12993#issuecomment-1247285900

**Comment by richvdh** [comment]
Yeah, there's a lot going on in that issue!

I've added some notes to the PR description. In general... it's super helpful if we can include a couple of notes about exactly what we're fixing and how in the PR description!

**Pull request merged**